inst/include/sampler.h

```diff
@@ -273,10 +274,21 @@ namespace smc {
         Moves = sFrom.Moves;
         /// The additional algorithm parameters.
         algParams = sFrom.algParams;
-        /// An object for adapting additional algorithm parameters
-        pAdapt = sFrom.pAdapt;
-        ///A flag to track whether the adaptation object needs to be included in this destructor.
-        adaptBelong = sFrom.adaptBelong;
+        if(sFrom.adaptBelong) {
+            // this can only happen if the default adaptMethods was used,
+            // i.e., no call to SetAdaptMethods
+            pAdapt = new adaptMethods<Space,Params>;
+            adaptBelong = 1;
+        } else {
+            // this can only happen if SetAdaptMethods was called,
+            // i.e., pAdapt points to an external object which should not be deleted with sampler
+            pAdapt = sFrom.pAdapt;
+            adaptBelong = 0;
```

Comment:
I'm not sure this is the right behaviour.

If we just copy a pointer like this, then we end up in a situation in which iterating either the original sampler or the new copy changes the configuration  of _both_ samplers which is clearly tricky.

It might not be completely trivial to resolve this in an elegant way; there was a reason that this constructor was absent prior to #28 and that was that it's a little delicate to do and not quite clear what the correct behaviour is. Perhaps this pointer to an object model wasn't such a good idea.

Reply:
Good point. I find this stuff quite difficult and I'm not sure how to handle it without either having slicing or dependence.

[This SO discussion](https://stackoverflow.com/questions/12255546/c-deep-copying-a-base-class-pointer) suggests doing a virtual function for the 'clone', but that seems annoying for users.

The suggestions in Bjarne Stroustrup's book seem to be centred around using clone functions too. The options he suggests are along the lines of preventing copying or "make the copy operations private or protected, and provide a clone() function to allow a user to request a non-slicing copy." I.e. using clone as above.

Reply:
Well, if we want to allow duplication of derived types from a base class I can't see a way of avoiding a virtual mechanism for doing it so unless there's some cunning way of doing this I'm not sure it can be avoided.

Perhaps it would make sense to agree that this PR at least addresses the memory-safeness (perhaps moduluo the explicit copy I suggested above) and add another item to the list to deal with allowing for proper deep-copies of these objects?

Reply:
Okay that seems like a good plan to me as well.

Reply:
Dear All, thanks for catching this.

Actually I was wondering why `SetAlgParam` passes by value whereas `SetAdaptMethods` does not, as the former does bypass this copying problem...